File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline

k, m = map(int, input().split())

lines = []
for _ in range(k):
  lines.append(int(input()))
lines.sort()

def binary_search(left, right, target):
  while left <= right:
    mid = (left + right) // 2
    count = 0
    for line in lines:  
      count += line // mid
    if count >= target:
      left = mid + 1
    else:
      right = mid - 1
  print(right)

binary_search(1, lines[-1], m)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : 'input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('\n');

[a, ...arr] = input;
[k, n] = a.split(' ').map((e) => Number(e));

arr = arr.map((e) => Number(e));
arr.sort((a, b) => a - b);

function solve(x) {
  let cur = 0;
  for (let i = 0; i < k; i++) {
    count = parseInt(arr[i] / x);
    cur += count;
  }

  if (cur >= n) {
    return true;
  } else return false;
}

start = 1;
end = arr[arr.length - 1];
while (start < end) {
  mid = parseInt((start + end + 1) / 2);

  if (solve(mid)) start = mid;
  else end = mid - 1;
}
console.log(start);
